Kernel/Term.hpp

```diff
@@ -966,21 +1016,54 @@ class Literal
   template<bool flip = false>
   unsigned hash() const
   {
-    bool positive = (flip ^ isPositive());
-    unsigned hash = DefaultHash::hash(positive ? (2*_functor) : (2*_functor+1));
-    if (isTwoVarEquality()) {
-      hash = HashUtils::combine(
-        DefaultHash::hash(twoVarEqSort()),
-        hash
-      );
+    return Literal::literalHash(functor(), polarity() ^ flip, 
+        [&](auto i) -> TermList const& { return *nthArgument(i); }, arity(),
+        someIf(isTwoVarEquality(), [&](){ return twoVarEqSort(); }),
+        commutative());
+  }
+
+  template<class GetArg>
+  static unsigned literalEquals(const Literal* lit, unsigned functor, bool polarity, GetArg getArg, unsigned arity, Option<TermList> twoVarEqSort, bool commutative) {
+    if (functor != lit->functor() || polarity != lit->polarity()) return false;
+
+    if (commutative) {
+      ASS_EQ(arity, 2)
+      ASS(rightArgOrder(getArg(0), getArg(1)))
+
+      if (someIf(lit->isTwoVarEquality(), [&](){ return lit->twoVarEqSort(); }) != twoVarEqSort) {
+        return false;
+      }
+      return std::make_tuple(*lit->nthArgument(0), *lit->nthArgument(1)) == std::make_tuple(getArg(0), getArg(1));
```

Comment:
What is the magic about make_tuple here? Is this a fancy way of saying 

`(*lit->nthArgument(0) == getArg(0)) && (*lit->nthArgument(1) == getArg(1))
`
or is more going on here?

Also, can we be sure that `rightArgOrder` also holds for `lit`?

Reply:
Yeah it's just a fancy way of saying that!
You're right we can't be sure but it's an implicit assumption i made. I added an assertion.